src/proxy/dns.rs

```diff
@@ -0,0 +1,271 @@
+    socket_addr.as_ref().parse().unwrap()
+}
+
+impl PollingDns {
+    pub(super) async fn new(pi: ProxyInputs, drain: Watch) -> Result<PollingDns, Error> {
+        info!(component = "dns", "dns async polling client started",);
+        Ok(PollingDns {
+            pi,
+            drain,
+            tasks: HashMap::new(),
+        })
+    }
+
+    fn get_handle(
+        proxy_mode: ProxyMode,
+        dns_nameservers: Vec<SocketAddr>,
+        state: Arc<RwLock<ProxyState>>,
+        dns_workload: Workload,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            let hostname = dns_workload.hostname.clone();
+            trace!("dns workload async task started for {:?}", &hostname);
+
+            // TODO(kdorosh): don't make a new forwarder for every request?
+            let fw = dns::forwarder_for_mode(proxy_mode, dns_nameservers).unwrap(); // TODO(kdorosh) handle unwrap
+            let r = fw.resolver();
+
+            // Lookup a host.
+            let req = a_request(
+                n(&hostname),
+                socket_addr("1.1.1.1:80"), // TODO(kdorosh): don't hardcode this
+                Protocol::Udp,
+            );
+            let resp = r.lookup(&req).await;
+            if resp.is_err() {
+                warn!(
+                    "dns async response for workload {} is: {:?}",
+                    &dns_workload.uid, resp
+                );
+                return;
+            } else {
+                trace!(
+                    "dns async response for workload {} is: {:?}",
+                    &dns_workload.uid,
+                    resp
+                );
+            }
+            let resp = resp.unwrap();
+            let ips = resp
+                .record_iter()
+                .filter_map(|record| {
+                    if record.rr_type().is_ip_addr() {
+                        // TODO: handle ipv6
+                        return record.data().unwrap().as_a().copied();
+                    }
+                    None
+                })
+                .collect_vec();
+            state
+                .write()
+                .unwrap()
+                .resolved_dns
+                .set_ips_for_workload(dns_workload.uid, ips);
+        })
+    }
+
+    pub(super) async fn run(mut self) {
+        let accept = async move {
+            loop {
+                let dns_workloads = self
+                    .pi
+                    .state
+                    .state
+                    .read()
+                    .unwrap()
+                    .workloads
+                    .get_async_dns_workloads();
+
+                // kill tasks that no longer need to be running
+                let current_workload_uids =
+                    dns_workloads.iter().map(|w| w.uid.clone()).collect_vec();
+                let current_workload_uid_set: HashSet<String> =
+                    HashSet::from_iter(current_workload_uids.iter().cloned());
```

Comment:
no need to make collect into current_workload_uids if we are then just going to re-collect it into a Set. We can go straight from iter -> Set

Reply:
addressed